### ISSUE TITLE
BUG: trying to upcast (Proc(Nil) | Nil) <- (Proc(String) | Nil) (Exception)

### ISSUE THREAD

**Issue by docelic** [crystal-lang/crystal#8428]
Here's code that causes a compiler exception:

```
def create_proc(b : String)
  
  if true
    begin
       ->(){""}
    rescue
      return ->(){}
    end
  end
end

create_proc "X"
```
(Also on https://play.crystal-lang.org/#/r/7y27 )

I was able to trigger this issue only in this somewhat narrow set of conditions -- an `if`, with a `begin` block, with the begin part not having `return`, while rescue do having it, and with the proc signature being different.

**Comment by rmarronnier** [comment]
Appears to be fixed : https://carc.in/#/r/990k

**Comment by jhass** [comment]
Let's add the example as a spec.

**Comment by docelic** [comment]
The issue reported in this ticket has been solved some 1.5 years ago.
The suggestion was to also add a test before closing this ticket, to make sure that the error doesn't come back at some later point.

But, I was unable to come up with a fully working test myself (#10073), and asking for help from others wasn't successful.
So, closing this ticket since obviously nothing else is going to happen.
